### PR TITLE
CB-21232 Add DB for EFM

### PR DESCRIPTION
**Pull request by bejancsaba**
See detailed description in the commit message.

### CODE REVIEW

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/efm/EfmRoleConfigProvider.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    static final String DRIVER_CLASS = "efm.db.driverClass";
+
+    @Override
+    protected List<ApiClusterTemplateConfig> getRoleConfigs(String roleType, TemplatePreparationObject source) {
+        RdsView efmRdsView = getRdsView(source);
+        List<ApiClusterTemplateConfig> configs = new ArrayList<>();
+
+        if (EfmRoles.EFM_SERVER.equals(roleType)) {
+            configs.add(config(DATABASE_URL, efmRdsView.getConnectionURL()));
```
Comment (lajosrodek) [inline template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/efm/EfmRoleConfigProvider.java:33]: Looks alright. Just a quick q: Which version of the PostgreSQL JDBC driver is being included / used? Asking for sake of SSL mode.

Context: When applicable, SSL mode options will be automatically included in `getConnectionURL()`, but the JDBC driver should be recent enough (>=42.2.5) in order to correctly interpret the DB server certificate chain. We raised a related bug ticket against NiFi Registry a couple of weeks ago; turned out they rely on some ancient JDBC driver that cannot cope with the Subject Alternative Name in the certificate chain. See [CFM-3228](https://jira.cloudera.com/browse/CFM-3228) if interested.

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/efm/EfmRoleConfigProvider.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    static final String DATABASE_URL = "efm.db.url";
+
+    static final String DRIVER_CLASS = "efm.db.driverClass";
+
+    @Override
+    protected List<ApiClusterTemplateConfig> getRoleConfigs(String roleType, TemplatePreparationObject source) {
+        RdsView efmRdsView = getRdsView(source);
+        List<ApiClusterTemplateConfig> configs = new ArrayList<>();
+
+        if (EfmRoles.EFM_SERVER.equals(roleType)) {
```
Comment (lajosrodek) [inline template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/efm/EfmRoleConfigProvider.java:32]: Does not hurt, but also superfluous. `getRoleTypes()` returns `List.of(EfmRoles.EFM_SERVER)`, which means this config provider will only be invoked & evaluated for the appropriate roles. Would only make a difference when multiple roles were supported. Anyway, can be left as is for now.

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/efm/EfmRoleConfigProvider.java
Comment (lajosrodek) [inline template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/efm/EfmRoleConfigProvider.java:1]: Please cover with UT.

**Review by lajosrodek — CHANGES_REQUESTED** [review]
Some minor asks.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/rdsconfig/EfmRdsConfigProvider.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    @Value("${cb.efm.database.db:efm}")
+    private String efmDb;
+
+    @Value("${cb.efm.database.port:5432}")
+    private String efmDbPort;
+
+    @Inject
+    private CmTemplateProcessorFactory cmTemplateProcessorFactory;
+
+    private boolean isRdsConfigNeedForEfm(Blueprint blueprint) {
```
[the remark this answers is not in the document]
Comment (bejancsaba) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/rdsconfig/EfmRdsConfigProvider.java]: Corrected

File: core/src/main/java/com/sequenceiq/cloudbreak/service/rdsconfig/EfmRdsConfigProvider.java
[the remark this answers is not in the document]
Comment (bejancsaba) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/rdsconfig/EfmRdsConfigProvider.java:1]: Done

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/efm/EfmRoleConfigProvider.java
Comment (bejancsaba) [reply template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/efm/EfmRoleConfigProvider.java:1]: Done

File: template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/efm/EfmRoleConfigProviderTest.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+@RunWith(MockitoJUnitRunner.class)
+public class EfmRoleConfigProviderTest {
+
+    private static final String CONNECTION_URL = "jdbc:postgresql://testhost:5432/efm";
+
+    private static final String USER_NAME = "efm_server_user";
+
+    private static final String USER_PASSWORD = "efm_server_password";
+
+    private final EfmRoleConfigProvider underTest = new EfmRoleConfigProvider();
```
Comment (lajosrodek) [inline template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/efm/EfmRoleConfigProviderTest.java:35]: Even if we expect the class to be stateless, it is safer to instantiate and assign `underTest` for each test case execution, i.e. using a `@BeforeEach setUp()` method (and also dropping `final` from the field).

A similar comment would also apply to `EfmRdsConfigProviderTest`.

**Review by lajosrodek — APPROVED** [review]
Thanks. Only two minor notes. Up to to if they get fixed, approving in advance nevertheless.

File: core/src/test/java/com/sequenceiq/cloudbreak/service/rdsconfig/EfmRdsConfigProviderTest.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+public class EfmRdsConfigProviderTest {
+
+    private static final String EFM = "efm";
+
+    private static final String PORT = "5432";
+
+    private final EfmRdsConfigProvider efmRdsConfigProvider = new EfmRdsConfigProvider();
+
+    @Test
+    public void testKnoxGatewayServiceRdsConfigProvider() {
```
[the remark this answers is not in the document]
Comment (bejancsaba) [reply core/src/test/java/com/sequenceiq/cloudbreak/service/rdsconfig/EfmRdsConfigProviderTest.java]: Renamed sorry copy paste :)

File: template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/efm/EfmRoleConfigProviderTest.java (hunk 4 above)
Comment (bejancsaba) [reply template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/efm/EfmRoleConfigProviderTest.java]: I made the changes you requested junit4 and junit5 is used in the two classes respectively so BeforeEach and Before was utilised.

**Comment by lajosrodek** [comment]
@bejancsaba Would you please squash your changes into a single commit?

**Review by lajosrodek — CHANGES_REQUESTED** [review]

File: template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/efm/EfmRoleConfigProviderTest.java (hunk 4 above)
Comment (lajosrodek) [reply template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/efm/EfmRoleConfigProviderTest.java]: JUnit 4 should be avoided; please see my below comment as well. :smile:

File: template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/efm/EfmRoleConfigProviderTest.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import static org.mockito.Mockito.when;
+
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.junit.MockitoJUnitRunner;
```
[the remark this answers is not in the document]
Comment (bejancsaba) [reply template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/efm/EfmRoleConfigProviderTest.java]: done

**Review by lajosrodek — APPROVED** [review]
Thanks, LGTM

**Pull request merged**